### ISSUE TITLE
macOS: Customize initial size for quick terminal

### ISSUE THREAD

**Issue by lucymhdavies** [ghostty-org/ghostty#2384]
In macOS, it looks like we're setting the default sizes here:
https://github.com/ghostty-org/ghostty/blob/main/macos/Sources/Features/QuickTerminal/QuickTerminalPosition.swift#L10-L29

For example, when showing at the top of the screen, we're setting the height to be 1/4 the height of the screen.
(specifically, this is whichever screen quick terminal first loads on. there's some edge cases around quick-terminal-screen=mouse if you have monitors of varying sizes, which I'm not going to go into right now)

I couldn't figure out from https://github.com/ghostty-org/ghostty/pull/2320/files how this is implemented on other systems. Am I missing something, or is this a macOS-only feature for now?

There are two options we could go for here:
* Allow the user to specify a multiplier in their config, which would default to 0.25
* Allow the user to specify a fixed height

(plus the option of both)

In my case, I would want my multiplier to be 1. i.e. the quick terminal takes up the entire screen. For my specific example, I see no value in specifying a fixed height. My instinct is that most people who use quick terminal have a preference for it to fill "a bit of the screen", "a lot of the screen" or "all of the screen". I imagine pixel-precision sizing of the quick terminal would be a niche desire.
(I could of course be wildly wrong these assumptions)

So were I to implement this myself (and I'm thinking I probably will), my proposal is:
* Add a new config item, e.g. `quick-terminal-size-multiplier`, with default value of 0.25
* Update where we set the window size, for example `height: screen.frame.height * multiplier`

There are a few edge cases I'd have to consider (e.g. making sure the multiplier is > 0, and <= 1), but otherwise this feels like it would be a fairly simple thing to implement.

If you're happy with the proposed implementation, I might get a PR raised for it over the weekend.

**Comment by sohsatoh** [comment]
This feature has been the biggest hurdle for me in fully replacing iTerm2 with Ghostty, so I’m currently working on implementing it.

It’s just a proof of concept for now, but the changes aren’t too significant.

If everything looks good, I’m going to submit a pull request in a few days (hopefully).

https://github.com/user-attachments/assets/1510f20d-5b53-4066-90d5-bd1aa8c3130e

**Comment by donaldguy** [comment]
I see the PR is in flight, so maybe preaching to the choir but

I would strongly appreciate some ability to dial in quick-terminal width when in top b/c

I use a 49" ultrawide (LG 49WL95C-W) and the ergonomics for `top` are pretty bad here:

![Image](https://github.com/user-attachments/assets/16bb122a-34f3-4e75-8262-41320affd3da)

(cf iTerm:

![Image](https://github.com/user-attachments/assets/263b9cbd-21ca-4273-93cb-060a66ba14d9)
)

and center is ... okay:

![Image](https://github.com/user-attachments/assets/1ae371e3-21c6-441f-86c7-412f8a9e64b8)


(but `top` IS much better than `center` when I unplug and am on the 13" display of the MacBook air that regularly drives that display; right now I deal with switching, but since its a requires-restart config, its not great)

[if anyone is wondering, Safari was being weird about previewing the image uploads; that's why it gets replaced by arc between screenshot 1 and 2)

**Comment by donaldguy** [comment]
Alas, having chosen to rebase and build first, read Swift later, I discover that even with the 2 number forms of `quick-terminal-size` that #3742 currently still hard-sticks `top` and `bottom` to full-width, e.g. here:

https://github.com/ghostty-org/ghostty/blob/8f5f432ab61fa73b77cb61239c0cff34250f77bb/macos/Sources/Features/QuickTerminal/QuickTerminalPosition.swift#L65-L69

(and the other 2 versions)

I suppose the ~problem is that [either you could still default to center OR] you need an additional config here (or a 4 number format?) to specify not only per-se width but also offset from edge

**Comment by donaldguy** [comment]
Umm, so ... I opened #4716 as a draft PR because it does contain a partial implementation that I probably plan to continue on. 

(and I originally was just gonna make a gist of a patch or indeed a <code>```diff</code> block in a comment on #3742, as I mostly wanted to give @dmehala et al a heads up I was working on an alternative. What would have been the rest of that comment more or less ended up as first comment: https://github.com/ghostty-org/ghostty/pull/4716#issuecomment-2573950501

but then I also got a little carried away and wrote, in order, 3 more lengthy comments:

- 1 one which perhaps would belong more in one or more issues: [Mouse resizing behavior](https://github.com/ghostty-org/ghostty/pull/4716#issuecomment-2573992931) 
    
- 2 which almost certainly really should be a discussion? - "Is this whole approach [to quick terminal (size)] wrong?": 
    - [(pt 1.)](https://github.com/ghostty-org/ghostty/pull/4716#issuecomment-2574032531) - contemplating whether "quick terminal size" is/ought-to-be handled in terms of a window, or rather more like a pane within a fullscreen (partially transparent) "window"
    - [(pt 2.)](https://github.com/ghostty-org/ghostty/pull/4716#issuecomment-2574068194) - contemplating wether adding any more than the existing (pretty specific) `quick-terminal-*` configs makes sense.

I… don't really think those are in the right place but I also don't know that they could be moved at this point and make sense (but mitchell or any other project admin types, feel free to do so anyway if you disagree)

I also think I'm burned out on thinking about this for the day or at least the next several hours, so … I wanted to go ahead and point these out to potentially interested parties before I give computing a break.

Cheers 😅

**Comment by abjugard** [comment]
It would be great to be able to specify size in "characters", e.g. old style rows/columns. Perhaps if specified without unit, e.g. `120, 32`?

**Comment by okuuva** [comment]
> It would be great to be able to specify size in "characters", e.g. old style rows/columns. Perhaps if specified without unit, e.g. `quick-terminal-size = 120, 32`?

I don't know if it's in any way feasible (or even a good idea), but I've found it intuitive what some neovim plugins do when defining dimensions: values <= 0 are proportional (percentage of the available space) while values > 0 are absolute (meaning rows or columns. The one (or at least the most) unintuitive thing about it is that fullscreen (either by width or height) means the value needs to be set to 0.

Making it so that floats would always be proportional and and ints absolute would fix that. But then it might be confusing that 30 and 30.0 might be completely different depending on the font size and the screen.

**Comment by afanjul** [comment]
Hi, is any of the "quick terminal height/size" feature already released? thanks

**Comment by avamsi** [comment]
Do I understand correctly that 100% will basically be fullscreen and will dynamically adjust to different display sizes (switching the quick terminal between the native display vs an external monitor, for example)?

**Comment by lmBored** [comment]
For now, a workaround I found it to set `window-save-state` to `always`, open quick terminal, use cursor to drag the bottom corner and expand the height of quick terminal to your liking, and that's it!

**Comment by avamsi** [comment]
@lmBored I don't think that works with multiple displays (of different sizes).

**Comment by lmBored** [comment]
> [@lmBored](https://github.com/lmBored) I don't think that works with multiple displays (of different sizes).

That's a good point, haven't tried it with 2 monitors yet, but resizing quick terminal manually is in pixel, so it should work in multiple displays in the sense that quick terminal size remains the same (instead of changing dynamically to scale with screen resolution)

**Comment by DefacedFace** [comment]
@lmBored I'll give you slurp frfr just ask

**Comment by lucymhdavies** [comment]
@lmBored 
> For now, a workaround I found it to set window-save-state to always, open quick terminal, use cursor to drag the bottom corner and expand the height of quick terminal to your liking, and that's it!

alas, that doesn't seem to persist across application restarts

**Comment by lmBored** [comment]
> [@lmBored](https://github.com/lmBored)
> 
> > For now, a workaround I found it to set window-save-state to always, open quick terminal, use cursor to drag the bottom corner and expand the height of quick terminal to your liking, and that's it!
> 
> alas, that doesn't seem to persist across application restarts

Fair point, I made this script to get the window_id of the quick terminal and resize it automatically with yabai. For usage, you can add something like `ctrl  + alt + shift + cmd - r: ~/script_name` to `.skhdrc` and press `hyper + r` to resize. It works just perfectly fine for me.

```bash
#!/usr/bin/env bash

# Constants
display_info=$(yabai -m query --displays)
GHOSTTY_WIDTH=$(echo "$display_info" | jq -r '.[0].frame.w') # 1470
GHOSTTY_HEIGHT=$(echo "$display_info" | jq -r '.[0].frame.h / 2') # 575
LOG_FILE="/tmp/ghostty_handler.log"

function log {
    echo "$(date): $1" >> "$LOG_FILE"
}

function ghostty_get_window_id {
    local windows
    local window_id
    
    windows=$(yabai -m query --windows)

    # Print all Ghostty windows and their titles
    echo "$windows" | jq -r '.[] | select(.app == "Ghostty") | "ID: \(.id), Title: \(.title)"' >> "$LOG_FILE"
    
    # Updated to match title ending with :~
    window_id=$(echo "$windows" | jq -r '.[] | select(.app == "Ghostty" and (.title | test(":~$|^~$")) ) | .id')
    
    if [[ -n "$window_id" ]]; then
        log "Found window ID: $window_id"
        echo -n "$window_id"
        return 0
    else
        log "No window ID found"
        return 1
    fi
}

function ghostty_resize {
    local window_id=$1
    
    if [[ -n "$window_id" ]]; then
        log "Attempting to resize window $window_id"
        yabai -m window "$window_id" --resize abs:$GHOSTTY_WIDTH:$GHOSTTY_HEIGHT
        log "Resized Ghostty window $window_id to ${GHOSTTY_WIDTH}x${GHOSTTY_HEIGHT}"
        return 0
    else
        log "No window ID to resize"
        return 1
    fi
}

function ghostty_manage {
    local window_id
    
    log "Starting ghostty_manage"
    if window_id=$(ghostty_get_window_id); then
        ghostty_resize "$window_id"
    else
        log "No matching Ghostty window found"
        return 1
    fi
}

ghostty_manage
```

**Comment by DoctorRyner** [comment]
> For now, a workaround I found it to set window-save-state to always, open quick terminal, use cursor to drag the bottom corner and expand the height of quick terminal to your liking, and that's it!

@lmBored umm, it doesn't work for me, the quick terminal does not resize at all

**Comment by pluiedev** [comment]
This is trivial to implement on Linux, but I do wonder how the proposed syntax would work for quick terminal placed in the center: primary/secondary axes don't really make sense in that instance.

**Comment by pluiedev** [comment]
Still needs to be implemented for macOS

**Comment by binaryben** [comment]
I don't know if this will matter for the already merged GTK changes, but for macOS, is there any reason the centred quick terminal option can't be a typical window that can be controlled by a WM like Rectangle? Or is it fixed for a reason? I have noticed that the resize cursor appears when hovering at the border of a quick terminal, but clicking and dragging does nothing.

I have a quake profile setup in iTerm that I just typically move where I want it with a Rectangle hotkey. Currently trying to replicate that in migrating to Ghostty

**Comment by abjugard** [comment]
Any progress on this? Jonesing to ditch iTerm2 for Ghostty, and this is the last blocker for me.

**Comment by dmehala** [comment]
Since GTK has been merged. I'll try to [revive my macOS PR](https://github.com/ghostty-org/ghostty/pull/3742).

**Comment by chrislockard** [comment]
Please also consider allowing customization of quick terminal width. It isn't obvious from this thread that width will be customizable in addition to height. This is useful for quake-style setups on wide/ultra-widescreen displays.

**Comment by donaldguy** [comment]
Quoting self from Jan 6: https://github.com/ghostty-org/ghostty/issues/2384#issuecomment-2574084132

>   "Is this whole approach [to quick terminal (size)] wrong?":
>   
>   * [(pt 1.)](https://github.com/ghostty-org/ghostty/pull/4716#issuecomment-2574032531) - contemplating whether "quick terminal size" is/ought-to-be handled in terms of a window, or rather more like a pane within a fullscreen (partially transparent) "window"

I have in fact taken to generally doing a runaround of:
1. uncommenting a `window-pad-x = 9050`, then 
2. reloading config <kbd>&#x2318;</kbd><kbd>,<kbd/>,
3. closing the quick terminal <kbd>&#x2303;&nbsp;Ctrl</kbd><kbd>D</kbd>, 
4. re-opening a new one with the config in effect ( <kbd>&#x2303;</kbd><kbd>`</kbd>)
5. re-commenting out the line ( <kbd>↑</kbd><kbd>&#x23CE;&nbsp;Return</kbd><kbd>R</kbd><kbd>#</kbd><kbd>:wq</kbd><kbd>&#x23CE;</kbd> )
6. re-re-loading the config for the benefit of other future windows <kbd>&#x2318;</kbd><kbd>,<kbd/>

its a stupid workaround, but it works (it also has the interesting emergent property that when I switch to laptop screen it automagically is correct also, I guess because the margins become invalid? [as 950*2 is probably > width on 13" air screen] or just that happens on main display change?

--- 

Anyway, idk if I had a point (I guess that I have not continued any more with my draft incomplete PR). I would certainly still welcome a real native solution for this. 

[whether as proper window resize or just the ability to short-circuit the workaround - in an ideal world like an immediate effect (vs requiring a new window, also ideally skipping the config file) `ghostty +do-local-config set-margin-x` 

(I went down the VT commands rabbit hole yesterday for other reasons, and tried today doing `\e[?69h\e[100;282s` and that actually _kinda_ works for my ultrawide management purposes, 

but it certainly gets fouled on other programs [notably frequently atuin] fouling the margins mode set, and—more surprisingly to me—the margin is also MIA after a ghostty level clear screen/scrollback <kbd>&#x2318;</kbd><kbd>K</kbd>) ]

**Comment by pluiedev** [comment]
> Please also consider allowing customization of quick terminal width. 

That is the secondary axis for top/bottom quick terminal and the primary one for left/right QTs

**Comment by freepicheep** [comment]
> Since GTK has been merged. I'll try to [revive my macOS PR](https://github.com/ghostty-org/ghostty/pull/3742).

I am interested in implementing this feature but don't want to take away from your work. Are you planning on reopening your PR and implementing this?

**Comment by dmehala** [comment]
Go for it @freepicheep :)

**Comment by kLy** [comment]
August now. Any updates?

**Comment by freepicheep** [comment]
We're waiting on Mitchell to review the changes on #7576. Reviews from anyone with Swift experience would help too, I'm sure.

**Comment by wilsonhou** [comment]
are there any workarounds for this in the meantime to resizing quick terminal?

**Comment by wilsonhou** [comment]
been checking this issue everyday and the absolute delight of seeing it merged is unmatched. thanks mitchell & @freepicheep !!!

**Comment by v3rm0n** [comment]
Does this mean we can do full screen quick terminal on macOS now? Been waiting for this to ditch iTerm2

**Comment by 0x5b62656e5d** [comment]
@v3rm0n Seems like the version with this feature will be released on September 15. There is a [nightly version](https://github.com/ghostty-org/ghostty/releases/tag/tip) you can try though.

**Comment by v3rm0n** [comment]
Thanks @0x5b62656e5d I'm on commit 6530107e3, configuration is valid, no errors, but the size is not doing anything when I try relative values with %, only px works:
```
auto-update-channel = tip

keybind = global:§=toggle_quick_terminal
quick-terminal-screen = mouse
quick-terminal-animation-duration = 0
quick-terminal-position = center
quick-terminal-size = 100%, 100%
```

Or do I need to do anything else to get the quick terminal to full screen?

**Comment by 0x5b62656e5d** [comment]
@v3rm0n Not too sure, you'll have to refer to other contributors/maintainers for this.
I'm still using stable release so I haven't tried out the feature yet.

**Comment by freepicheep** [comment]
Hey @v3rm0n, I'm not entirely sure why it's not working on tip. I'm guessing it's because some of the checks were not successful on the merge into main.

<img width="878" height="516" alt="Image" src="https://github.com/user-attachments/assets/5c87ad5e-5f7e-46d4-8d2a-96b8a0f3f925" />

**Comment by mitchellh** [comment]
Unrelated @freepicheep. It's in the tip release. Looks like there might be some other bugs, trying to reproduce now. #8419

**Comment by v3rm0n** [comment]
Oh yeah it does work on one of my computers but not on the other, weird.

**Comment by mitchellh** [comment]
Please open new discussions for issues from here on out.
